doug-martin/nestjs-query
# Custom filter support for typeorm and graphql

Hey! In the past few days I spent some time thinking about a general solution for having custom user-defined filters ( related issues: https://github.com/doug-martin/nestjs-query/issues/702 and https://github.com/doug-martin/nestjs-query/issues/720 ), for now only at the typeorm level.
I'm opening this draft PR to ask for feedback on my solution. If this can work for you, I can probably keep working on this and transform it in a real PR.

**Problem:** It is currently not possible to define user-defined filters on entities. This means that, for example, if we're using postgis and our entity has a Point field (e.g. GPS coordinates) it's impossible to create a filter that returns all entities that are closer than X to a given point (radius search), since the only allowed operations are the ones defined by the sqlComparisonBuilder.

Assuming our entity has a `location` field which represents its position, I'd like to be able to filter on it with something like `{location: {distanceFrom: {point: {lat: ..., lng: ...}, radius: ....}}}`.

**Proposed solution:** 
Extend the filtering mechanism to allow custom user defined filters that can be applied during the transformation of the `Query` object into the actual typeorm query.
A custom filter is basically the following interface: 
```ts
export type CustomFilterResult = { sql: string; params: ObjectLiteral };

export interface CustomFilter<Entity = unknown, T = unknown> {
  apply(field: keyof Entity & string, cmp: string, val: T, alias?: string): CustomFilterResult;
}
```
(note that CustomFilterResult is the same as CmpSQLType https://github.com/luca-nardelli/nestjs-query/blob/custom-filters/packages/query-typeorm/src/query/sql-comparison.builder.ts#L9)

Each custom filter can be registered for a specific entity class, field and operation name, and is then invoked by the where builder when it's processing the filter object. Here, "operation name" basically means the individual operation of the filter (same as `cmpType` in the where builder)
```
{ numberType: { gt: 10 } }
```
In the above example, `numberType` is the field, and `gt` is the operation name. Filter registration happens through a new `CustomFilterRegistry` class that basically holds all custom filters for all entities in a Map. Storing them flat allows to easily fetch all custom filters when traversing relations in the where builder. 

I started writing some code with the above idea in mind and I've managed to cobble together something that allowed me to write and pass some tests ( added here https://github.com/luca-nardelli/nestjs-query/blob/custom-filters/packages/query-typeorm/__tests__/query/where.builder.spec.ts).

More specifically, this is how it looks like at the code level for 2 different custom filters: an extension for numeric types where I want to get values that are multiples of a user provided value, and an example with a PostGIS query.

```ts

  const customFilterRegistry = new CustomFilterRegistry();
  customFilterRegistry.setFilter<TestEntity>(TestEntity, 'numberType', 'isMultipleOf', {
    apply(field, cmp, val: number, alias): CustomFilterResult {
      alias = alias ? alias : '';
      const pname = `param${randomString()}`;
      return {
        sql: `("${alias}"."${field}" % :${pname}) == 0`,
        params: { [pname]: val },
      };
    },
  });

  customFilterRegistry.setFilter<TestEntity>(TestEntity, 'fakePointType', 'distanceFrom', {
    apply(field, cmp, val: { point: { lat: number; lng: number }; radius: number }, alias): CustomFilterResult {
      alias = alias ? alias : '';
      const plat = `param${randomString()}`;
      const plng = `param${randomString()}`;
      const prad = `param${randomString()}`;
      return {
        sql: `ST_Distance("${alias}"."${field}", ST_MakePoint(:${plat},:${plng})) <= :${prad}`,
        params: { [plat]: val.point.lat, [plng]: val.point.lng, [prad]: val.radius },
      };
    },
  });

  const expectSQLSnapshot = (filter: Filter<TestEntity>): void => {
    const selectQueryBuilder = createWhereBuilder().build(
      getQueryBuilder(),
      filter,
      {},
      TestEntity, // <--- New parameter
      customFilterRegistry, // <--- New parameter
      'TestEntity',
    );
    const [sql, params] = selectQueryBuilder.getQueryAndParameters();
    expect(sql).toMatchSnapshot();
    expect(params).toMatchSnapshot();
  };

  it('should accept custom filters alongside regular filters', (): void => {
    expectSQLSnapshot({
      numberType: { gte: 1, lte: 10, isMultipleOf: 5 },
      fakePointType: { distanceFrom: { point: { lat: 45.3, lng: 9.5 }, radius: 50000 } },
    } as any);
  });
```
The custom filter registry here is a standard class and I am registering filters manually, but at a higher level we could register custom filters using decorators + DI. Apart from this, the most notable changes I had to make so far are the following:
- Add an explicit reference to the class currently being processed to the build() method of the where builder and to all of its inner methods.
- Add an explicit reference to the custom filter registry to the build() method of the where builder and to all of its inner methods.
- Rewrite the getRelationNamesRecursive method to not only extract relation names, but also the target class, in order to use that as index for the custom filter registry.

At the wherebuilder level, the main change happened in the `withFilterComparison`, where now I try to call a custom filter and I use the sqlcomparisonbuilder only as a fallback.

```ts
const sqlComparisons = opts.map((cmpType) => {
          const customFilter = customFilterRegistry?.getFilter(klass, field, cmpType);
          // If we have a registered customfilter for this cmpType, this has priority over the standard sqlComparisonBuilder
          if (customFilter) {
            return customFilter.apply(field, cmpType, cmp[cmpType], alias);
          }
          return this.sqlComparisonBuilder.build(
            field,
            cmpType,
            cmp[cmpType] as EntityComparisonField<Entity, T>,
            alias,
          );
        });
        sqlComparisons.map(({ sql, params }) => qb.orWhere(sql, params));
```


-----

There are obviously a lot of things to do here (as an example, there's some type fixing that needs to happen because right now we are assuming that the only possible values in the `Filter*` types are the common ones), but for now could you @doug-martin (or someone else) have a look and tell me if you think this is an acceptable approach?

Thanks!
